modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteServerImpl.java

```diff
@@ -188,13 +236,64 @@ public CompletableFuture<Void> waitForInitAsync() {
         return joinFuture;
     }
 
+    /**
+     * Restarts the node asynchronously. The {@link Ignite} instance obtained via {@link #api()} and objects acquired through it remain
+     * functional, but completion of calls to them might be delayed during the restart (that is, synchronous calls might take more time,
+     * while futures from asynchronous calls might take more time to complete).
+     *
+     * @return CompletableFuture that gets completed when the node startup has completed (either successfully or with an error).
+     */
+    CompletableFuture<Void> restartAsync() {
```

Comment:
It looks like there might be a race here.

While we started restarting the node, it is shutdown in the process and the first checks that are indicated here will already be passed.
What about using not mutex but CAS on the `restartOrShutdownFuture` and each new operation will be performed after the previous one is completed and will thus be linearized and with the correct checks.

Reply:
I fixed it a bit differently: I added `shutDown` flag that is rechecked on restart after acquiring the mutex.
The future chaining is already implemented.
CAS logic seems to be a bit more involved compared to simple and rude `synchronized`

Reply:
It looks more complicated to me, but it's up to you.